File: loop_transformations/sources/apply_split.1.c
/*
* @@name:       apply_split.1
* @@type:       C
* @@operation:  compile
* @@expect:     success
* @@version:    omp_6.0
*/
float compute(int i){return (float)i; }

void split_composable(int m, int n, float A[n])
{
   #pragma omp split counts(m,omp_fill) \
       apply(split: target loop nowait map(from:A[0:m]), \
                    parallel loop)
   for (int i = 0; i < n; ++i)
      A[i] = compute(i);

   #pragma omp taskwait
}

void split_composable_equivalent(int m, int n, float A[n])
{
    #pragma omp target loop nowait map(from:A[0:m])
    for (int i = 0; i < m; ++i)
       A[i] = compute(i);

    #pragma omp parallel loop
    for (int i = m; i < n; ++i)
       A[i] = compute(i);

   #pragma omp taskwait
}

File: loop_transformations/sources/apply_split.1.f90
! @@name:       apply_split.1
! @@type:       F-free
! @@operation:  compile
! @@expect:     success
! @@version:    omp_6.0
function compute(i) result(computed)
  integer :: i
  real :: computed
  computed=real(i)
end function

subroutine split_composable(m, n, A)
   implicit none
   integer, intent(in ) :: m, n
   real,    intent(out) :: A(n)
   integer :: i
   real,external :: compute

   !$omp  split counts(m,omp_fill)  &
   !$omp& apply(split: target loop nowait map(from:A(1:m)), &
   !$omp&              parallel loop)
   do i=1,n
      A(i) = compute(i);
   end do

   !$omp taskwait

end subroutine

subroutine split_composable_equivalent(m, n, A)
   implicit none
   integer, intent(in ) :: m, n
   real,    intent(out) :: A(n)
   integer :: i
   real,external :: compute

  !$omp target loop nowait map(from:A(1:m))
  do i = 1,m
    A(i) = compute(i)
  end do

  !$omp parallel loop
  do i = m+1,n
    A(i) = compute(i)
  end do

  !$omp taskwait

end subroutine
